### PR TITLE
Calibration screen

### LINKED ISSUE #103: Calibration/Debug Screen
Extend `dpsctl.py -c` to allow setting the DACs like so `dpsctl.py -c VOUT_DAC=10 `

This would be a useful addition to working towards completing #17

### PR DESCRIPTION
**Pull request by Xenoamor**
This should hopefully close #103

The settings ui currently comprises of a single screen called calibration which has a cross-hair for an icon. The layout is as follows:
```
V_DAC    V_OUT_ADC
A_DAC    A_OUT_ADC
         V_IN_ADC
```
This will be clearer once we have a full alphabet available to use.

You can switch to the screen by either pressing both M1 and M2 or by issuing the command: `python dpsctl.py --screen settings` you may switch back to the main screen with `python dpsctl.py --screen main`

All the `-F -f -P -p -q` options work from this screen

This will also be a useful framework for once we start implementing actual settings on the device

### CODE REVIEW

**Comment by Xenoamor** [comment]
Please hold on this until #116 is merged so I can use an alphabet to label the values

File: dpsctl/dpsctl.py [hunk 1]
```diff
@@ -471,6 +473,14 @@ def handle_commands(args):
     if args.calibration_reset:
         communicate(comms, create_cmd(cmd_clear_calibration), args)
 
+    if args.switch_screen:
+        if (args.switch_screen.lower() == "main"):
+            communicate(comms, create_change_screen(change_screen_main), args)
+        elif (args.switch_screen.lower() == "settings"):
+            communicate(comms, create_change_screen(change_screen_settings), args)
+        else:
+            fail("please specify either settings or main as parameters")
```
Comment (kanflo) [inline dpsctl/dpsctl.py]: Perhaps "...either 'settings' or 'main' ..." for clarity?

File: dpsctl/dpsctl.py [hunk 2]
```diff
@@ -661,6 +671,7 @@ def main():
     parser.add_argument('-v', '--verbose', action='store_true', help="Verbose communications")
     parser.add_argument('-V', '--version', action='store_true', help="Get firmware version information")
     parser.add_argument('-U', '--upgrade', type=str, dest="firmware", help="Perform upgrade of OpenDPS firmware")
+    parser.add_argument(      '--screen', type=str, dest="switch_screen", help="Switch to or from the settings screen")
```
Comment (kanflo) [inline dpsctl/dpsctl.py]: I think the two options should be mentioned here.

File: opendps/hw.c [hunk 3]
```diff
@@ -696,9 +719,23 @@ void BUTTON_M1_isr(void)
...
         exti_set_trigger(BUTTON_M1_EXTI, EXTI_TRIGGER_RISING);
     } else {
-        event_put(event_button_m1, press_short);
+        m1_pressed = false;
+
+        if (!m2_pressed) {
+            if (m1_and_m2_pressed) {
+                m1_and_m2_pressed = false;
+                event_put(event_buttom_m1_and_m2, press_short);
+            }
```
Comment (kanflo) [inline opendps/hw.c]: May I be picky about braces? ;)

File: opendps/opendps.c [hunk 4]
```diff
@@ -715,6 +735,31 @@ void opendps_upgrade_start(void)
...
+    uui_disable_cur_screen(current_ui); /** Turn off the output */
+    opendps_update_power_status(false); /** Update the power icon status */
+
+    if (screen_id == FUNC_UI_ID)
+        current_ui = &func_ui;
+    else if (screen_id == SETTINGS_UI_ID)
+        current_ui = &settings_ui;
+
+    tft_clear(); /** Clear any previous screen */
+    uui_activate(current_ui);
```
Comment (kanflo) [inline opendps/opendps.c]: Note to self: at some point clearing should be moved inside `uui_activate`.

**Comment by kanflo** [comment]
Cool! Looks good.

**Comment by kanflo** [comment]
Great. I will revisit this next week, got a few other things to sort out.

**Comment by Xenoamor** [comment]
Thanks @tzarc! This is really great, I'll see what I can do with it when I get a chance

File: opendps/hw.c (hunk 3 above)
Comment (Xenoamor) [reply opendps/hw.c]: How do you prefer the braces? Allman style?

Comment (kanflo) [reply opendps/hw.c]: I am actually a bit of an K&R guy.

**Comment by Xenoamor** [comment]
Okay this is all good to go I believe. Next step after this will finally be kerbing #17 
image

File: opendps/settings_calibration.c [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+/**
+ * @brief      Callback for when the function is enabled
+ *
+ * @param[in]  enabled  true when function is enabled
+ */
+static void calibration_enable(bool enabled)
+{
+    emu_printf("[Calibration] %s output\n", enabled ? "Enable" : "Disable");
+    if (enabled) {
+        pwrctl_set_ilimit(10000);
```
Comment (kanflo) [inline opendps/settings_calibration.c]: Is this magic number intended? ;)

**Comment by kanflo** [comment]
Looks nice! Should we perhaps go with SI units and say `Iout` instead of `Aout`? Also, I think the printed ADC and DAC values should be defined as `.unit = unit_none,` since they represent, but are not a, voltage/current.

File: opendps/settings_calibration.c (hunk 5 above)
Comment (Xenoamor) [reply opendps/settings_calibration.c]: Ooops, This it to stop the OCP firing. I've added a comment to it to clarify

**Comment by Xenoamor** [comment]
That makes sense to me!
image

**Comment by tzarc** [comment]
Looks good! Yeah the font is better IRL, photos seem to bleed a bit.

**Pull request merged**